OpenSSL/Session.hsc

```diff
@@ -392,6 +393,25 @@ contextSetALPNProtos context protos =
       -- https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_alpn_protos.html#RETURN-VALUES
       _ssl_set_alpn_protos ctx cFormattedProtos (fromIntegral len) >>= failIf_ (/= 0)
 
+type KeylogCb = Ptr SSL_ -> CString -> IO ()
+
+foreign import ccall "wrapper" mkKeylogCb :: KeylogCb -> IO (FunPtr KeylogCb)
+
+foreign import capi "openssl/ssl.h SSL_CTX_set_keylog_callback" _ssl_ctx_set_keylog_callback :: Ptr SSLContext_ -> FunPtr KeylogCb -> IO ()
+
+-- | The key logging callback is called with a String "line". The line is a
+-- string containing the key material in the format used by NSS for its
+-- SSLKEYLOGFILE debugging output. To recreate that file, the key logging
+-- callback should log line, followed by a newline.
+contextSetKeylogCallback :: SSLContext -> (String -> IO ()) -> IO ()
+contextSetKeylogCallback context cb =
+  withContext context $ \ctx -> do
+    -- There doesn't seem to be a way to go from 'Ptr SSL_' to 'SSL', so let's
+    -- just ignore it in the haskell callback.
+    cbPtr <- mkKeylogCb $ \_ssl line ->
```

Comment:
`FunPtr` requires [freeHaskellFunPtr](https://hackage.haskell.org/package/base-4.18.0.0/docs/Foreign-Ptr.html#v:freeHaskellFunPtr). Perhaps it worth to change this function to `withContextSetKeylogCallback` which will remove the callback and free its memory on exit.